### PR TITLE
ENH: Sieve variable support

### PR DESCRIPTION
**Pull request by monoidic**
This fixes issue #2486

### CODE REVIEW

**Comment by sebix** [comment]
Wow, very neat again!

@kamil-certat as you have a use-case of this feature, do you want to have a look at the diff?

File: intelmq/bots/experts/sieve/expert.py [hunk 1]
```diff
@@ -41,219 +40,309 @@
...
-        'Expression': lambda self, match, event: self.match_expression(match, event),
+    _date_op_map = {":before": operator.lt, ":after": operator.gt}
+
+    _cond_map: Dict[
+        str,
+        Callable[
+            [
+                "SieveModel",
+                object,
+                Event,
```
Comment (kamil-certat) [inline intelmq/bots/experts/sieve/expert.py]: NIT: it could be any _Message_. Although it's rare, it is possible to process Report object as well

File: intelmq/bots/experts/sieve/expert.py [hunk 2]
```diff
@@ -41,219 +40,309 @@
...
+        for path in paths:
+            self.send_message(event, path=path)
+
+        self.acknowledge_message()
+
+
+class SieveModel:
+    def __init__(self, file_path: str, logger) -> None:
+        self.metamodel = self.init_metamodel()
+        self.sieve = self.read_sieve_file(file_path, self.metamodel)
```
Comment (kamil-certat) [inline intelmq/bots/experts/sieve/expert.py]: NIT: I'd ask to rename `sieve` here or in the bot class - just for sake of confusing, those are two variables with the same name, relatively close to each other, and with a different meaning. Maybe name it in the bot as `sieve_model`?

File: intelmq/bots/experts/sieve/expert.py [hunk 3]
```diff
@@ -263,48 +352,40 @@ def process_conjunction(self, conj, event) -> bool:
 
     def process_condition(self, cond, event) -> bool:
         name = cond.match.__class__.__name__
-        return cond.neg ^ self._cond_map[name](self, cond.match, event)
+        ret = self._cond_map[name](self, cond.match, event)
+        if cond.neg:
+            ret = not ret
+        return ret
```
Comment (kamil-certat) [inline intelmq/bots/experts/sieve/expert.py]: Did you decide to rewrite because of readability, or there is a change I don't see? For sake of readability, I'd think about `return not ret if cond.neg else ret`. In cases like this I find it a little clearer to keep this in a one line. But it's personal preference.

File: intelmq/bots/experts/sieve/expert.py [hunk 4]
```diff
@@ -263,48 +352,40 @@ def process_conjunction(self, conj, event) -> bool:
...
-            return op in {'!=', '!~'}
+            return op in {"!=", "!~"}
 
-        lhs = event[key]
-        if not isinstance(lhs, str) and op not in ('==', '!='):
-            if isinstance(lhs, dict):
-                lhs = dumps(lhs)
-            else:
-                lhs = str(lhs)
+        return self._string_op_map[op](event[key], value)
```
Comment (kamil-certat) [inline intelmq/bots/experts/sieve/expert.py]: I'd ask for a short explanation why:
* `process_single_string_match`
* `process_multi_string_match`
* `process_multi_numeric_match`
were changed. 

The `value(s).value(s)` -> `value(s)` change looks clear, it's been moved upper to the operator maps, but I'm unsure about dropped conversion and special cases for numeric match, where is it handled now? (I might have missed it)

File: intelmq/bots/experts/sieve/expert.py [hunk 5]
```diff
@@ -419,33 +511,33 @@ def validate_ip_range(ip_range) -> None:
...
         """
-        valid_types = {'Integer', 'Float', 'Accuracy', 'ASN'}
-        position = SieveExpertBot.get_linecol(num_match.value, as_dict=True)
+        valid_types = {"Integer", "Float", "Accuracy", "ASN"}
+        position = SieveModel.get_linecol(num_match.value, as_dict=True)
 
         # validate harmonization type (event key)
         try:
-            type = SieveExpertBot._harmonization[num_match.key]['type']
+            type = SieveExpertBot._harmonization[num_match.key]["type"]
```
Comment (kamil-certat) [inline intelmq/bots/experts/sieve/expert.py:532]: I'd think about explicit setting the harmonisation in the `SieveModel` class. Now it's a little strange that the `SieveModel` class, just used by `SieveExpertBot`, then silently have a backref to `SieveExpertBot` - and thus to a value that by default is `None`. If you try to draw how classes depend on each other, I would be surprised seeing a circle between classes not inheriting on each other.

File: intelmq/tests/bots/experts/sieve/test_expert.py [hunk 6]
```diff
@@ -1719,33 +1854,41 @@ def test_mixed_if_and_actions(self):
...
-        expected['extra.list_empty'] = True
+        expected["extra.list_empty"] = True
 
         self.input_message = event
         self.run_bot()
         self.assertMessageEqual(0, expected)
 
-    def test_extra_dict(self):
-        self.sysconfig['file'] = os.path.join(os.path.dirname(__file__), 'test_sieve_files/test_extra_dict.sieve')
+    def test_variables(self):
```
Comment (kamil-certat) [inline intelmq/tests/bots/experts/sieve/test_expert.py:1888]: (!) What has happened to the`test_extra_dict` test case?

**Review by kamil-certat — CHANGES_REQUESTED** [review]
I love the change, as well as the refactoring and formatting against black/ruff. I have some NIT comments and one or two clarification questions, but also a one big one: what has happened to the test `test_extra_dict`? Have you found it duplicated (if so, please remove test files and say what duplicates it)?

Please also update the bot documentation to describe variables as well as the changelog.

And as a discussion, I have two things I'm unsure, but I don't find it blocking:
1) a variable can be in place of typed values, like `Bool`. Shouldn't we then ensure in code that the variable has the correct value?
2) variables look like extremely promising for the future, but I also ask myself - wouldn't it be also nice just allow syntax like `ifd.field1 = idf.field2`? The current use case looks like only supporting this.
3) I'd think about extending the validation - I haven't seen if there is any place where we would try to statically validate that the requested variable was previously defined

File: intelmq/bots/experts/sieve/expert.py (hunk 4 above)
Comment (monoidic) [reply intelmq/bots/experts/sieve/expert.py]: The removed code in `process_multi_numeric_match` was after an unconditional `return` and hence completely dead code.
The rest of the changes are in order to uniformly add `resolve_value` to all match functions via `_cond_map`, to allow for the use of variables in those matches.

The changes in `process_single_string_match` appear to have been due to me starting development from an older commit and accidentally discarding the newer commits.

File: intelmq/bots/experts/sieve/expert.py (hunk 5 above)
Comment (monoidic) [reply intelmq/bots/experts/sieve/expert.py:532]: Perhaps splitting the bot in two was not a great idea. I re-merged them, though kept the previous `SieveModel.process` separate from `SieveExpertBot.process` as `model_process`

File: intelmq/tests/bots/experts/sieve/test_expert.py (hunk 6 above)
Comment (monoidic) [reply intelmq/tests/bots/experts/sieve/test_expert.py:1888]: Accidentally discarded the commi which added this.

**Comment by monoidic** [comment]
> what has happened to the test test_extra_dict?

Accidentally discarded newer commits to the Sieve bot. I noticed it in the commit diff ~a week ago, but got distracted and forgot to address it...

> a variable can be in place of typed values, like Bool. Shouldn't we then ensure in code that the variable has the correct value?

This has actually been addressed in the code. The `resolve_value` method which resolves variables (if needed) has a parameter, `expected_types`, expecting a type or tuple thereof. If specified, it raises an exception if the variable's type is not as expected.
Or do you mean that an expected type could somehow be specified in the Sieve file itself as well?

> variables look like extremely promising for the future, but I also ask myself - wouldn't it be also nice just allow syntax like `ifd.field1 = idf.field2`? The current use case looks like only supporting this.

At the moment, yes.
The parsing end would technically accept both `append extra.list $elem` as well as `append $list $elem` at the moment (with the former appending a value given in a variable to a list in the event, and the latter working purely with variables), though the processing end does not handle variables as keys yet.
Allowing event keys to be used like that could be an option, but seems like a slight downgrade feature-wise (e.g any temporary variables which are not meant to end up in the event itself would have to exist under random keys of the event and be cleared out afterwards) while requiring all the same new validation that's been added for variables anyway.
Though for some cases, it would be more ergonomical.

> I'd think about extending the validation - I haven't seen if there is any place where we would try to statically validate that the requested variable was previously defined

I'd expect `check()` to be the place for this.

**Review by kamil-certat — APPROVED** [review]
Thanks for the explanation! It sounds sufficient for me.

Before it gets merged, I'd ask you to verify / add a test that `check()` would reject a sieve file with variable used without being declared beforehand

**Comment by sebix** [comment]
Merging this, based on @kamil-certat's positive review.

For the last request on a check for undeclared variables, I opened https://github.com/certtools/intelmq/issues/2526
